tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.ElevatedServer/QuietBackgroundProcessesSession.cpp

```diff
@@ -0,0 +1,134 @@
+    public:
+        STDMETHODIMP RuntimeClassInitialize() noexcept
+        {
+            return S_OK;
+        }
+
+        // IQuietBackgroundProcessesSession
+        STDMETHODIMP Start(__int64* result) noexcept override try
+        {
+            auto lock = std::scoped_lock(g_mutex);
+
+            // Stop and discard the previous timer
+            if (g_activeTimer)
+            {
+                g_activeTimer->Cancel();
+            }
+
+            std::chrono::seconds duration = DEFAULT_QUIET_DURATION;
+            if (auto durationOverride = try_get_registry_value_dword(HKEY_LOCAL_MACHINE, LR"(Software\Microsoft\Windows\CurrentVersion\DevHome\QuietBackgroundProcesses)", L"Duration"))
+            {
+                duration = std::chrono::seconds(durationOverride.value());
+            }
+
+            // Start timer
+            g_activeTimer.reset(new TimedQuietSession(duration));
+
+            // Return duration for showing countdown
+            *result = g_activeTimer->TimeLeftInSeconds();
+            return S_OK;
+        }
+        CATCH_RETURN()
+
+        STDMETHODIMP Stop() noexcept override try
+        {
+            auto lock = std::scoped_lock(g_mutex);
+
+            // Turn off quiet mode and cancel timer
+            if (g_activeTimer)
+            {
+                g_activeTimer->Cancel();
+                g_activeTimer.reset();
+            }
+
+            return S_OK;
+        }
+        CATCH_RETURN()
+
+        STDMETHODIMP get_IsActive(::boolean* value) noexcept override try
+        {
+            auto lock = std::scoped_lock(g_mutex);
+            *value = false;
+            if (g_activeTimer)
+            {
+                *value = g_activeTimer->IsActive();
+            }
+            return S_OK;
+        }
+        CATCH_RETURN()
+
+        STDMETHODIMP get_TimeLeftInSeconds(__int64* value) noexcept override try
+        {
+            auto lock = std::scoped_lock(g_mutex);
+            *value = 0;
+            if (g_activeTimer)
+            {
+                *value = g_activeTimer->TimeLeftInSeconds();
+            }
+            return S_OK;
+        }
+        CATCH_RETURN()
+    };
+
+    class QuietBackgroundProcessesSessionStatics WrlFinal :
+        public Microsoft::WRL::AgileActivationFactory<
+            Microsoft::WRL::Implements<IQuietBackgroundProcessesSessionStatics>>
+    {
+        InspectableClassStatic(RuntimeClass_DevHome_QuietBackgroundProcesses_QuietBackgroundProcessesSession, BaseTrust);
+
+    public:
+        STDMETHODIMP ActivateInstance(_COM_Outptr_ IInspectable**) noexcept
+        {
+            // Disallow activation - must use GetSingleton()
+            return E_NOTIMPL;
+        }
+
+        // IQuietBackgroundProcessesSessionStatics
+        STDMETHODIMP GetSingleton(_COM_Outptr_ IQuietBackgroundProcessesSession** session) noexcept override try
+        {
+            THROW_IF_FAILED(Microsoft::WRL::MakeAndInitialize<QuietBackgroundProcessesSession>(session));
```

Comment:
>MakeAndInitialize [](http://example.com/codeflow?start=44&length=17)

Why does GetSingleton spin up a new instance on every call?

Reply:
I realize that the *client* attempts to ensure singleton-ness, but IMO if all of the work is being done client-side, the server might as well just use the normal ActivateInstance entrypoint; a server GetSingleton method implies that the server is the one guaranteeing singleton instancing.

Reply:
I'd say even if server guaranteeing singleton instancing it can return it from ActivateInstance. Unless both instancing models need to be supperted I'd use the normal ActivateInstance entrypoint.

Reply:
Agree. I don't care between GetSingleton and ActivateInstance - GetSingleton makes a bit more sense to me, but the QuietBackgroundProcessesSession object is already semantically skewed when the TimedQuietSession was introduced.

Reply:
Okay I don't get it. What's the magic to implement this without keeping the server alive?  I tried implementing with weak references but they actually increase the refcount of COM in this process for some severely broken reason.

        // IQuietBackgroundProcessesSessionStatics
        STDMETHODIMP GetSingleton(_COM_Outptr_ IQuietBackgroundProcessesSession** session) noexcept override try
        {
            if (m_weakSingleton)
            {
                if (auto strong = m_weakSingleton.query<IQuietBackgroundProcessesSession>())
                {
                    *session = strong.detach();
                    return S_OK;
                }
            }

            wil::com_ptr<IQuietBackgroundProcessesSession> sessionInstance;
            THROW_IF_FAILED(Microsoft::WRL::MakeAndInitialize<QuietBackgroundProcessesSession>(&sessionInstance));
            m_weakSingleton = wil::com_weak_query(sessionInstance);
            *session = sessionInstance.detach();
            return S_OK;
        }
        CATCH_RETURN()

Reply:
Yeah, singleton + server shuts down when idle are tough to reconcile. If the singleton keeps the server alive, that's obviously a violation. If the singleton doesn't keep the server alive, you have to deal with "somebody tries to access the singleton while the server is in the middle of shutting down" scenarios.

As discussed offline, "instanced objects with shared state" may be more practical than a true out-of-proc singleton.

Reply:
Yep.  Instanced objects is the only feasible way I can think of to move forward.  An 'instance' can be thought of as a 'handle' to 'the singleton' backend.